stream_alert/rule_processor/sink.py

```diff
@@ -92,3 +95,88 @@ def sink(self, alerts):
+                                'RuleDescription': {'S': alert['rule_description']},
+                                'Outputs': {'SS': alert['outputs']},
+                                # Compact JSON encoding (no extra spaces)
+                                'Record': {'S': json.dumps(alert['record'], separators=(',', ':'))}
+                            }
+                        }
+                    }
+                    for alert in batch
+                ]
+            }
+
+    @backoff.on_exception(backoff.expo, ClientError, max_tries=5, jitter=backoff.full_jitter,
+                          on_backoff=backoff_handlers.backoff_handler,
+                          on_success=backoff_handlers.success_handler,
+                          on_giveup=backoff_handlers.giveup_handler)
+    def _batch_write(self, request_items, max_attempts=5):
+        """Write a batch of alerts to Dynamo, retrying with exponential backoff for failed items.
+
+        Args:
+            request_items (dict): The generated RequestItems dict from _alert_batches().
+            max_attempts (int): Maximum number of times to retry UnprocessedItems.
+
+        Returns:
+            (bool) True if the batch write was eventually successful, False otherwise.
+        """
+        for attempt in xrange(1, max_attempts + 1):
+            response = self.client_dynamo.batch_write_item(RequestItems=request_items)
+
+            # If Dynamo experiences an internal error, unprocessed items are listed in the response.
+            # AWS recommends retrying unprocessed items in a loop with exponential backoff.
+            request_items = response['UnprocessedItems']
+            if request_items:
+                LOGGER.warn(
+                    'Batch write failed: %d alerts were not written (attempt %d/%d)',
+                    len(request_items[self.table]), attempt, max_attempts)
+                # Simple exponential backoff: Sleep 0.5, 1, 2, 4 and 8 seconds.
+                time.sleep(0.25 * 2 ** attempt)
+                attempt += 1
+            else:
+                return True
+
+        return False
+
+    def _sink_dynamo(self, alerts):
+        """Write alerts in batches to Dynamo."""
+        for batch_num, batch in enumerate(self._alert_batches(alerts), start=1):
+            LOGGER.info('Sending batch #%d to Dynamo with %d alert(s)',
+                        batch_num, len(batch[self.table]))
+            if not self._batch_write(batch):
+                LOGGER.error('Unable to save alert batch %s', json.dumps(batch))
+
+    def sink(self, alerts):
+        """Sink triggered alerts from the StreamRules engine.
+
+        Args:
+            alerts (list): A list of dictionaries representing json alerts.
+        """
+        self._sink_lambda(alerts)
+
+        # While we are testing this, exceptions should be logged but not raise errors.
+        try:
+            self._sink_dynamo(alerts)
+        except Exception:  # pylint: disable=broad-except
```

Comment:
Is this a loose `except` for now just so we don't bring the whole ship down? Can I assume this will be refined after testing, etc? If so I'd add a TODO so we don't forget

Reply:
Yeah, this is to just prevent the rule processor from failing because of a problem sending to Dynamo. I added a TODO to remove/refine it after testing